numpy/core/tests/test_numeric.py

```diff
@@ -3602,3 +3602,12 @@ def test_zero_dimensional(self):
         arr_0d = np.array(1)
         ret = np.tensordot(arr_0d, arr_0d, ([], []))  # contracting no axes is well defined
         assert_array_equal(ret, arr_0d)
+
+    def test_subclass(self):
+        class Subclass(np.ndarray):
+            pass
+        a = np.ones((1, 2)).view(Subclass)
+        b = np.ones((2, 3)).view(Subclass)
+        td = np.tensordot(a, b, 1)
+        assert isinstance(td, Subclass)
+        assert_array_equal(td, np.dot(a, b))
```

Comment:
@InessaPawson pointed this out as one of the first-time contributor PRs that has been stalled for a bit (sorry!).

I'll summarize some analysis below, largely based on the comment from @seberg over in gh-21958.

It sounds like Sebastian had some concern that we'd break cases where users fed in `np.matrix` to `tensordot`, because on `main`/stable releases that currently converts the output to an `ndarray` rather than preserving the `np.matrix` type, and changing behavior like that in NumPy can be risky business/break a bunch of stuff. On the flip side, `np.matrix` is already documented as discouraged in favor of using regular arrays, so I'm not sure how much weight the preservation of its behavior should have long-term.

If we wanted to pursue a conversative route, I wonder what folks might think about this slight adjustment that allows preservation of the old behavior for `matrix` but empowers subclass pass-through otherwise:

```diff
--- a/numpy/core/numeric.py
+++ b/numpy/core/numeric.py
@@ -1076,7 +1076,12 @@ def tensordot(a, b, axes=2):
         axes_b = [axes_b]
         nb = 1
 
-    a, b = asanyarray(a), asanyarray(b)
+
+    if isinstance(a, np.matrix) or isinstance(b, np.matrix):
+        a, b = asarray(a), asarray(b)
+    else:
+        a, b = asanyarray(a), asanyarray(b)
+
     as_ = a.shape
     nda = a.ndim
     bs = b.shape
diff --git a/numpy/core/tests/test_numeric.py b/numpy/core/tests/test_numeric.py
index 89c920c1d..ad24495f6 100644
--- a/numpy/core/tests/test_numeric.py
+++ b/numpy/core/tests/test_numeric.py
@@ -3615,3 +3615,8 @@ class Subclass(np.ndarray):
         td = np.tensordot(a, b, 1)
         assert isinstance(td, Subclass)
         assert_array_equal(td, np.dot(a, b))
+        mat = np.matrix('1 2; 3 4')
+        result = np.tensordot(mat, mat, 1)
+        # preserve behavior from `main` -- don't pass
+        # through for matrix type
+        assert not isinstance(result, np.matrix)
```

It feels a bit awkward to bend over backwards like this for `matrix` at this point--I'm sure other folks will have opinions.

The final thing that needs to be decided is Sebastian's point about whether this should just be delegated to `__array_function__`, which is one of our mechanisms for downstream folks to override default behavior for their custom subtypes. It sounds like the drawback there may be that this is largely impractical if you only wanted to override for a single function, unless that has changed recently?

Beyond that, I did check that the rebased version of this PR passes the full test suite locally both with and without the diff above.

Hope it helps reach a decision here a bit more quickly!

Reply:
I think this would also impact `MaskedArray` which also inherits from `np.ndarray`. In a more perfect world, both these classes would already implement `__array_function__`, but for now I don't think we can safely change behaviour by using `asanyarray` here.